srv/modules/runners/disk_led.py

```diff
@@ -0,0 +1,208 @@
+    because it is mostly not possible to collect this data on storage
+    nodes due to missing privileges.
+    :return: Returns the OSD metadata.
+    :rtype: list
+    """
+    __master_opts__ = salt.config.client_config("/etc/salt/master")
+    __master_utils__ = salt.loader.utils(__master_opts__)
+    __salt_master__ = salt.loader.minion_mods(
+        __master_opts__, utils=__master_utils__)
+    master_minion = __salt_master__['master.minion']()
+
+    local_client = salt.client.LocalClient()
+    return local_client.cmd(master_minion, 'osd.metadata', tgt_type='compound')
+
+
+def osd(osd_id, device_type, led, status):
+    """
+    Manage the storage enclosure LED by OSD identifier.
+    :param osd_id: The OSD id, e.g.
+        - 4
+        - osd.1
+    :type osd_id: str
+
+    :param device_type: The device to use. This can be:
+        - primary (Blue-/Filestore)
+        - wal (Bluestore)
+        - db (Bluestore)
+        - journal (Filestore)
+        - all
+    :type device_type: str
+    :param led: The LED to be processed. This can be 'ident' or 'fault'.
+    :type led: str
+    :param status: Set to 'on' to switch on the light, otherwise 'off'.
+    :type status: str
+    """
+    osd_id = int(re.sub(r'^osd\.', '', str(osd_id)))
+    local_client = salt.client.LocalClient()
+
+    osd_metadata = _get_metadata()
+
+    # Get OSD information from all storage nodes.
+    osd_details = local_client.cmd(
+        'I@roles:storage', 'osd.details', [osd_metadata], tgt_type='compound')
+    # Process OSD information.
+    # {
+    #   "admin.ceph": {},
+    #   "mon1.ceph": {},
+    #   "data1.ceph": [{
+    #     "cluster": "ceph",
+    #     "id": 5,
+    #     "name": "osd.5",
+    #     "hostname": "data1",
+    #     "type": "bluestore",
+    #     "primary": {
+    #       "rotational": true,
+    #       "device_name": "vdg",
+    #       "device_file": "/dev/vdg"
+    #     },
+    #     "wal": {
+    #       "rotational": null,
+    #       "device_name": null,
+    #       "device_file": null
+    #     },
+    #     "db": {
+    #       "rotational": null,
+    #       "device_name": null,
+    #       "device_file": null
+    #     },
+    #     "journal": {
+    #       "rotational": null,
+    #       "device_name": null,
+    #       "device_file": null
+    #     },
+    #   }]
+    # }
+    for hostname in osd_details:
+        filtered_osd_details = list(
+            [x for x in osd_details[hostname] if x['id'] == osd_id])
+        log.debug("hostname: {} osds: {}".format(
+            hostname, pprint.pformat(filtered_osd_details)))
+        if not filtered_osd_details:
+            continue
+        osd_info = filtered_osd_details[0]
+        device_types = [device_type]
+        if device_type == 'all':
+            device_types = ['primary', 'wal', 'db', 'journal']
+        for dev_type in device_types:
+            device_name = osd_info[dev_type]['device_name']
```

Comment:
device_name is in the format of '/dev/dm-0'. (at least on my cluster, deployed with vagrant-ceph)

in the `_process(..)` function you are prepending the device_name with '/dev/' which results in:

`'/dev//dev/dm-0'`

Reply:
Can you please post the output of ``ceph osd metadata``?

Reply:
```
snip..
    {
        "id": 35,
        "arch": "x86_64",
        "back_addr": "[v2:172.16.2.21:6816/38848,v1:172.16.2.21:6817/38848]",
        "back_iface": "eth3",
        "bluefs": "1",
        "bluefs_single_shared_device": "1",
        "bluestore_bdev_access_mode": "blk",
        "bluestore_bdev_block_size": "4096",
        "bluestore_bdev_dev_node": "/dev/dm-0",
        "bluestore_bdev_driver": "KernelDevice",
        "bluestore_bdev_partition_path": "/dev/dm-0",
        "bluestore_bdev_rotational": "1",
        "bluestore_bdev_size": "20401094656",
        "bluestore_bdev_support_discard": "0",
        "bluestore_bdev_type": "hdd",
        "ceph_release": "nautilus",
        "ceph_version": "ceph version 14.0.1-3346-g0364ec8d2f (0364ec8d2f8783d21a7153abd3bf67361230622e) nautilus (dev)",
        "ceph_version_short": "14.0.1-3346-g0364ec8d2f",
        "cpu": "Intel Core Processor (Haswell, no TSX, IBRS)",
        "default_device_class": "hdd",
        "device_ids": "",
        "devices": "vdb",
        "distro": "sles",
        "distro_description": "SUSE Linux Enterprise Server 15 SP1",
        "distro_version": "15.1",
        "front_addr": "[v2:172.16.1.21:6816/38848,v1:172.16.1.21:6817/38848]",
        "front_iface": "eth2",
        "hb_back_addr": "[v2:172.16.2.21:6818/38848,v1:172.16.2.21:6819/38848]",
        "hb_front_addr": "[v2:172.16.1.21:6818/38848,v1:172.16.1.21:6819/38848]",
        "hostname": "data1",
        "journal_rotational": "1",
        "kernel_description": "#1 SMP Thu Feb 14 12:19:57 UTC 2019 (e1536c0)",
        "kernel_version": "4.12.14-32-default",
        "mem_swap_kb": "0",
        "mem_total_kb": "1516264",
        "network_numa_unknown_ifaces": "eth2,eth3",
        "objectstore_numa_unknown_devices": "vdb",
        "os": "Linux",
        "osd_data": "/var/lib/ceph/osd/ceph-35",
        "osd_objectstore": "bluestore",
        "rotational": "1"
    }
..snip
```

Reply:
Same problem here:

```
# salt-run disk_led.osd 1 all ident on
Turning on the ident LED of disk "/dev//dev/dm-0" on host "data4.ceph" ...
Exception occurred in runner disk_led.osd: Traceback (most recent call last):
  File "/usr/lib/python3.6/site-packages/salt/client/mixins.py", line 387, in _low
    data['return'] = self.functions[fun](*args, **kwargs)
  File "/srv/modules/runners/disk_led.py", line 158, in osd
    _process(hostname, device_name, led, status)
  File "/srv/modules/runners/disk_led.py", line 37, in _process
    _cmd_run(hostname, cmd_cfg[led][status].format(device_file=device_file))
  File "/srv/modules/runners/disk_led.py", line 51, in _cmd_run
    hostname, result[hostname]['ret']))
RuntimeError: Failed to execute command on "data4.ceph": NOT_FOUND_DISK(209): Disk /dev//dev/dm-0 not found
```

Looks like `blue_*_dev_node` is already the full device path.

Reply:
Oh great, they changed the output format. OK, need to adapt this.